File: 87_Problem_2.py
# code_report Solution
# https://youtu.be/K_KrVZnagAU

def longestMountain (self, A):
   up, dn, ans = [0] * len(A), [0] * len(A), 0

   for i in range(1, len(A)): 
      if A[i] > A[i - 1]: up[i] = up[i - 1] + 1

   for i in range(len(A) - 2, 0, -1):
      if A[i] > A[i + 1]: dn[i] = dn[i + 1] + 1

   for i in range(0, len(A)):
      if up[i] and dn[i]: ans = max (ans, up[i] + dn[i] + 1)

   return ans

File: 87_Problem_2.cpp
// code_report Solution
// https://youtu.be/K_KrVZnagAU

#include <vector>
#include <algorithm>

using namespace std;

int longestMountain (vector<int> A) 
{
   int n = A.size (), ans = 0;
   vector<int> up (n), dn (n);
      
   for (int i = n - 2; i >= 0; --i) if (A[i] > A[i + 1]) dn[i] = dn[i + 1] + 1;
   for (int i = 1; i < n; ++i)      if (A[i] > A[i - 1]) up[i] = up[i - 1] + 1;
   for (int i = 0; i < n; ++i)      if (up[i] && dn[i]) ans = max (ans, up[i] + dn[i] + 1);
      
   return ans;
}
